io/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -463,20 +462,22 @@ private object RTS {
                       }
                     } else {
                       // Error caught:
-                      val value = -\/(error)
+                      // FIXME this cast is not safe, `err` is defined as `E => IO[E2, B]`
+                      val handledIo = errorHandler(error).asInstanceOf[IO[E, Any]]
 
                       if (finalizer eq null) {
                         // No finalizer to run:
-                        curIo = nextInstr[E](value, stack)
+                        curIo = handledIo
 
-                        if (curIo eq null) {
-                          eval = false
-                          result = ExitResult.Completed(value)
-                        }
+                        // This will never happen. Right?
+                        // if (curIo eq null) {
```

Comment:
If the exception is caught, you always safely have a "next action", namely, the one created by feeding the error to the error handler.